File: 0102_binaryTreeLevelOrderTraversal/main.py
from typing import Optional, List
import collections

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        queue:collections.deque[Optional[TreeNode]] = collections.deque()
        queue.append(root)
        res = []
        while queue:
            level = []
            for i in range(len(queue)):
                curNode = queue.popleft()
                level.append(curNode.val)
                if curNode.left:
                    queue.append(curNode.left)
                if curNode.right:
                    queue.append(curNode.right)
            res.append(level)

        return res

File: 0102_binaryTreeLevelOrderTraversal/main.cpp
#include <queue>
#include <vector>
#include <utility>

using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // BFS Implementation
    vector<vector<int>> levelOrder(TreeNode* root) {

        if(!root)
            return std::vector<vector<int>>();

        std::vector<std::vector<int>> res;

        std::queue<struct TreeNode*> myQueue;
        myQueue.push(root);

        struct TreeNode *curPtr = nullptr;
        int levelLength = 0;

        while(!myQueue.empty()) {
            levelLength = myQueue.size();
            res.push_back(std::vector<int>());
            for(int i = 0; i < levelLength; ++i) {
                curPtr = myQueue.front();
                myQueue.pop();
                res.back().push_back(curPtr->val);
                if(curPtr->left)
                    myQueue.push(curPtr->left);
                if(curPtr->right)
                    myQueue.push(curPtr->right);
            }
        }
        return res;
    }
};
